lib/supavisor/protocol/client.ex

```diff
@@ -0,0 +1,157 @@
+defmodule Supavisor.Protocol.Client do
+  require Logger
+
+  @pkt_header_size 5
+
+  defmodule(Pkt,
+    do: defstruct([:tag, :len, :payload, :bin])
+  )
+
+  def pkt_header_size, do: @pkt_header_size
+
+  def header(<<char::8, pkt_len::32>>) do
+    {tag(char), pkt_len}
+  end
+
+  @spec decode(binary) :: {:ok, [Pkt.t()], binary} | {:error, any}
+  def decode(data) do
+    decode(data, [])
+  end
+
+  @spec decode(binary, [Pkt.t()]) :: {:ok, [Pkt.t()], binary} | {:error, any}
+  def decode("", acc), do: {:ok, Enum.reverse(acc), ""}
+
+  def decode(data, acc) do
+    case decode_pkt(data) do
+      {:ok, pkt, rest} -> decode(rest, [pkt | acc])
+      {:error, :payload_too_small} -> {:ok, Enum.reverse(acc), data}
+    end
+  end
+
+  @spec decode_pkt(binary) ::
+          {:ok, Pkt.t(), binary}
+          | {:acc, nil, binary}
+          | {:error, :payload_too_small}
+  def decode_pkt(<<_::8, pkt_len::32, payload::binary>>)
+      when byte_size(payload) < pkt_len - 4 do
+    {:error, :payload_too_small}
+  end
+
+  def decode_pkt(<<char::8, pkt_len::32, rest::binary>>) do
+    case tag(char) do
+      nil ->
+        {:error, {:undefined_tag, <<char>>}}
+
+      tag ->
+        payload_len = pkt_len - 4
+        <<bin_payload::binary-size(payload_len), rest2::binary>> = rest
+
+        {:ok,
+         %Pkt{
+           tag: tag,
+           len: pkt_len + 1,
+           payload: decode_payload(tag, bin_payload),
+           bin: <<char, pkt_len::32, bin_payload::binary>>
+         }, rest2}
+    end
+  end
+
+  def decode_pkt(_), do: {:error, :header_mismatch}
+
+  @spec tag(byte) :: atom | nil
+  def tag(char) do
+    case char do
+      ?Q ->
+        :simple_query
+
+      ?H ->
+        :flush_message
+
+      ?P ->
+        :parse_message
+
+      ?B ->
+        :bind_message
+
+      ?D ->
+        :describe_message
+
+      ?E ->
+        :execute_message
+
+      ?S ->
+        :sync_message
+
+      ?X ->
+        :termination_message
+
+      _ ->
+        nil
+    end
+  end
+
+  def decode_payload(:simple_query, payload) do
+    case String.split(payload, <<0>>) do
+      [query, ""] -> query
+      _ -> :undefined
```

Comment:
What happens when you pass `:undefined` or `nil` to Rust, which is expecting a string?

Reply:
It will raise the exception:
```elixir
iex(node1@127.0.0.1)3> Supavisor.PgParser.statement_types nil
** (ArgumentError) argument error
    (supavisor 0.9.5) Supavisor.PgParser.statement_types(nil)
```
Perhaps in that case, I should send the query to the write replica and not crash the process